gymnasium/experimental/functional_jax_env.py

```diff
@@ -0,0 +1,258 @@
+
+    def reset(self, *, seed: int | None = None, options: dict | None = None):
+        """Resets the environment."""
+        super().reset(seed=seed)
+        if seed is not None:
+            self.rng = jrng.PRNGKey(seed)
+
+        rng, self.rng = jrng.split(self.rng)
+
+        rng = jrng.split(rng, self.num_envs)
+
+        self.state = self.func_env.initial(rng=rng)
+        obs = self.func_env.observation(self.state)
+        info = self.func_env.state_info(self.state)
+
+        self.steps = jnp.zeros(self.num_envs, dtype=jnp.int32)
+
+        obs = _convert_jax_to_numpy(obs)
+
+        return obs, info
+
+    def step(self, action: ActType):
+        """Steps through the environment using the action."""
+        if self._is_box_action_space:
+            assert isinstance(self.action_space, gym.spaces.Box)  # For typing
+            action = np.clip(action, self.action_space.low, self.action_space.high)
+        else:  # Discrete
+            # For now we assume jax envs don't use complex spaces
+            assert self.action_space.contains(
+                action
+            ), f"{action!r} ({type(action)}) invalid"
+
+        rng, self.rng = jrng.split(self.rng)
+
+        rng = jrng.split(rng, self.num_envs)
+
+        next_state = self.func_env.transition(self.state, action, rng)
+        reward = self.func_env.reward(self.state, action, next_state)
+        terminated = self.func_env.terminal(next_state)
+
+        self.steps += 1
+
+        truncated = jnp.zeros_like(terminated)
+        info = self.func_env.step_info(self.state, action, next_state)
+        self.state = next_state
+
+        done = jnp.logical_or(terminated, truncated)
+        if jnp.any(done):
+            # TODO: record the final observations and return them in the info, check the format
+            to_reset = jnp.where(done)[0]
+            reset_count = to_reset.shape[0]
+            rng, self.rng = jrng.split(self.rng)
+            rng = jrng.split(rng, reset_count)
+            new_initials = self.func_env.initial(rng)
+            self.state = self.state.at[to_reset].set(new_initials)
+            self.steps = self.steps.at[to_reset].set(0)
+
+        observation = self.func_env.observation(next_state)
+        observation = _convert_jax_to_numpy(observation)
+
+        return observation, reward, terminated, truncated, info
+
+    def render(self):
+        """Returns the render state if `render_mode` is "rgb_array"."""
+        if self.render_mode == "rgb_array":
+            self.render_state, image = self.func_env.render_image(
+                self.state, self.render_state
+            )
+            return image
+        else:
+            raise NotImplementedError
+
+    def close(self):
+        """Closes the environments and render state if set."""
+        if self.render_state is not None:
+            self.func_env.render_close(self.render_state)
+            self.render_state = None
+
+
+def _convert_jax_to_numpy(element: Any):
```

Comment:
Can we not use the `experimental.wrappers.jax_to_numpy` functions?

Reply:
I think using the wrapper will be tricky at this moment, since the environment would have to be registered with a default wrapper, which isn't a thing we support at the moment afaik. Without the wrapper being applied, CI freaks out because it expects everything to be in numpy, which we should probably change as we get more jax stuff. But I should be able to use the function defined in that file.

Reply:
The functions for the wrapper are separate so we should be able to use the functions without the wrapper.

Reply:
Done